markets/perps-market/test/integration/Position/InterestRate.tolerance.test.ts

```diff
@@ -0,0 +1,125 @@
+    bootstrapMarkets({
+      interestRateParams: {
+        lowUtilGradient: interestRateParams.lowUtilGradient.toBN(),
+        gradientBreakpoint: interestRateParams.gradientBreakpoint.toBN(),
+        highUtilGradient: interestRateParams.highUtilGradient.toBN(),
+      },
+      synthMarkets: [],
+      perpsMarkets: [
+        {
+          lockedOiRatioD18: _ETH_LOCKED_OI_RATIO.toBN(),
+          requestedMarketId: 25,
+          name: 'Ether',
+          token: 'snxETH',
+          price: _ETH_PRICE.toBN(),
+        },
+        {
+          lockedOiRatioD18: _BTC_LOCKED_OI_RATIO.toBN(),
+          requestedMarketId: 50,
+          name: 'Bitcoin',
+          token: 'snxBTC',
+          price: _BTC_PRICE.toBN(),
+        },
+      ],
+      traderAccountIds: [2, 3],
+    });
+
+  let ethMarket: PerpsMarket, btcMarket: PerpsMarket;
+
+  before('setup', async () => {
+    ethMarket = perpsMarkets()[0];
+    btcMarket = perpsMarkets()[1];
+
+    // add collateral to margin
+    await systems().PerpsMarket.connect(trader1()).modifyCollateral(2, 0, bn(50_000));
+    await systems().PerpsMarket.connect(trader2()).modifyCollateral(3, 0, bn(200_000));
+
+    // set monthly price
+    await ethMarket.aggregator().mockSetMonthlyTolerancePrice(_ETH_MONTHLY_PRICE.toBN());
+    await btcMarket.aggregator().mockSetMonthlyTolerancePrice(_BTC_MONTHLY_PRICE.toBN());
+  });
+
+  before('positions open', async () => {
+    await openPosition({
+      systems,
+      provider,
+      trader: trader1(),
+      accountId: 2,
+      keeper: keeper(),
+      marketId: ethMarket.marketId(),
+      sizeDelta: _TRADER_ETH_SIZE.toBN(),
+      settlementStrategyId: ethMarket.strategyId(),
+      price: _ETH_PRICE.toBN(),
+    });
+
+    await openPosition({
+      systems,
+      provider,
+      trader: trader2(),
+      accountId: 3,
+      keeper: keeper(),
+      marketId: btcMarket.marketId(),
+      sizeDelta: _TRADER_BTC_SIZE.toBN(),
+      settlementStrategyId: btcMarket.strategyId(),
+      price: _BTC_PRICE.toBN(),
+    });
+  });
+
+  const checkMarketInterestRate = (withMontlyTolerance: boolean) => {
+    let currentInterestRate: Wei;
+    it('has correct interest rate', async () => {
+      const withdrawableUsd = wei(await systems().Core.getWithdrawableMarketUsd(superMarketId()));
+      const totalCollateralValue = wei(await systems().PerpsMarket.totalGlobalCollateralValue());
+      const delegatedCollateral = withdrawableUsd.sub(totalCollateralValue);
+
+      const minCredit = calculateMinCredit(withMontlyTolerance);
+
+      const utilRate = minCredit.div(delegatedCollateral);
+      currentInterestRate = calculateInterestRate(utilRate, interestRateParams);
+      const actualInterestRate = await systems().PerpsMarket.interestRate();
+      assertBn.near(actualInterestRate, currentInterestRate.toBN(), bn(0.0001));
+    });
+
+    return {
+      currentInterestRate: () => currentInterestRate,
+    };
+  };
+
+  checkMarketInterestRate(true);
```

Comment:
When abstracting code in tests please abstract simple functions that do not contain mocha hooks like before/describe/it and only have them on a top level.